provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type jujuConfigProvider struct {
+	key            []byte
+	keyFingerprint string
+	passphrase     string
+	tenancyOCID    string
+	userOCID       string
+	region         string
+}
+
+type ociClient struct {
+	ociCore.ComputeClient
+	ociCore.BlockstorageClient
+	ociCore.VirtualNetworkClient
+	ociIdentity.IdentityClient
+
+	ociCommon.ConfigurationProvider
+}
+
+// NewJujuConfigProvider returns a new ociCommon.ConfigurationProvider instance
+func NewJujuConfigProvider(user, tenant string, key []byte, fingerprint, passphrase, region string) ociCommon.ConfigurationProvider {
+	return &jujuConfigProvider{
+		key:            key,
+		keyFingerprint: fingerprint,
+		passphrase:     passphrase,
+		tenancyOCID:    tenant,
+		userOCID:       user,
+		region:         region,
+	}
+}
+
+func NewOciClient(provider ociCommon.ConfigurationProvider) (ApiClient, error) {
```

Comment:
We try to follow the pattern where Acronyms are all fully capitalized. (eg NewHTTPClient rather than NewHttpClient).

So this would be NewOCIClient returning an APIClient